popper/constrain.py

```diff
@@ -53,63 +42,37 @@ def vo_variable(variable):
 
 class Constrain:
     def __init__(self, experiment):
-        # self.no_pruning  = experiment.args.no_pruning
-        self.included_clause_handles = set()
         self.seen_clause_handle = {}
+        self.added_programs = set()
+        self.added_clauses = set()
 
     def literal_handle(self, literal):
         return f'{literal.predicate}{"".join(literal.arguments)}'
 
     def make_clause_handle(self, clause):
-        if clause not in self.seen_clause_handle:
-            body_literals = sorted(clause.body, key = operator.attrgetter('predicate'))
-            clause_handle = ''.join(self.literal_handle(literal) for literal in [clause.head] + body_literals)
-            self.seen_clause_handle[clause] = clause_handle
-        return self.seen_clause_handle[clause]
+        if clause in self.seen_clause_handle:
+            return self.seen_clause_handle[clause]
+        body_literals = sorted(clause.body, key = operator.attrgetter('predicate'))
+        clause_handle = ''.join(self.literal_handle(literal) for literal in [clause.head] + body_literals)
+        self.seen_clause_handle[clause] = clause_handle
+        return clause_handle
 
-    # AC: add caching
     def make_program_handle(self, program):
         return f'prog_{"_".join(sorted(self.make_clause_handle(clause) for clause in program.clauses))}'
 
-    def build_constraints(self, program_outcomes):
-        for program, (positive_outcome, negative_outcome) in program_outcomes:
-            constraint_types = self.derive_constraint_types(positive_outcome, negative_outcome)
-            for constraint in self.derive_constraints(program, constraint_types):
-                yield constraint
-            for inclusion_rule in self.derive_inclusion_rules(program, constraint_types):
-                yield inclusion_rule
-
-    def derive_constraint_types(self, positive_outcome, negative_outcome):
-        # if self.no_pruning:
-            # positive_outcome = Outcome.ALL
-            # negative_outcome = Outcome.NONE
-        # AC: @RM, what is this line about?
-        if negative_outcome == Outcome.ALL:
-            negative_outcome = Outcome.SOME
```

Comment:
If you don't use these two lines you need another three entries in the `OUTCOME_TO_CONSTRAINTS` table (one for every positive outcome combined with negative outcome `ALL`.